src/MongoDB.Driver.Core/Core/Bindings/CoreSession.cs

```diff
@@ -37,8 +38,9 @@ public sealed class CoreSession : ICoreSession
         private bool _isCommitTransactionInProgress;
         private readonly IOperationClock _operationClock = new OperationClock();
         private readonly CoreSessionOptions _options;
+        private IServer _pinnedShardRouter;
```

Comment:
It there an edge case where making this an `IServer` and not a `DnsEndpoint` could cause a problem?  Is there ever a situation in a sharded cluster where the IServer instance changes, especially with SRV monitoring?

It seems really unlikely, but possible.  @rstam?